File: src/2958.cpp
#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
   public:
    int maxSubarrayLength(vector<int>& nums, int k) {
        int retVal = 0;

        int numsSize = nums.size();
        unordered_map<int, int> frequency;
        int start = -1;
        for (int end = 0; end < numsSize; ++end) {
            frequency[nums[end]]++;
            while (frequency[nums[end]] > k) {
                start++;
                frequency[nums[start]]--;
            }
            retVal = max(retVal, (end - start));
        }

        return retVal;
    }
};

int main(int argc, char** argv) {
    struct subject {
        vector<int> nums;
        int k;
    };
    vector<subject> testData{
        {{1, 2, 3, 1, 2, 3, 1, 2}, 2}, {{1, 2, 1, 2, 1, 2, 1, 2}, 1}, {{5, 5, 5, 5, 5, 5, 5}, 4}, {{1, 4, 4, 3}, 1}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: nums = [1,2,3,1,2,3,1,2], k = 2
     *  Output: 6
     *
     *  Input: nums = [1,2,1,2,1,2,1,2], k = 1
     *  Output: 2
     *
     *  Input: nums = [5,5,5,5,5,5,5], k = 4
     *  Output: 4
     *
     *  Input: nums = [1,4,4,3], k = 1
     *  Output: 2
     */

    Solution cSolution;
    int answer;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: nums = [";
        for (long unsigned int j = 0; j < testData[i].nums.size(); ++j) {
            cout << ((j == 0) ? "" : ",") << testData[i].nums[j];
        }
        cout << "], k = " << testData[i].k << "\n";

        answer = cSolution.maxSubarrayLength(testData[i].nums, testData[i].k);
        cout << "Output: " << answer << "\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}

File: src/2958.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        retVal = 0

        frequency = defaultdict(int)
        start = -1
        for end, value in enumerate(nums):
            frequency[value] += 1
            while frequency[value] > k:
                start += 1
                frequency[nums[start]] -= 1
            retVal = max(retVal, end - start)

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for nums, k in zip([[1, 2, 3, 1, 2, 3, 1, 2], [1, 2, 1, 2, 1, 2, 1, 2], [5, 5, 5, 5, 5, 5, 5], [1, 4, 4, 3]],
                           [2, 1, 4, 1]):
            # /* Example
            #  *  Input: nums = [1,2,3,1,2,3,1,2], k = 2
            #  *  Output: 6
            #  *
            #  *  Input: nums = [1,2,1,2,1,2,1,2], k = 1
            #  *  Output: 2
            #  *
            #  *  Input: nums = [5,5,5,5,5,5,5], k = 4
            #  *  Output: 4
            #  *
            #  *  Input: nums = [1,4,4,3], k = 1
            #  *  Output: 2
            #  */
            logging.info("Input: nums = %s, k = %s", nums, k)

            retVal = pSolution.maxSubarrayLength(nums, k)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
